### PR TITLE
Support policy format change in v3.7.0 - #671 (Replaces #674)

### LINKED ISSUE #671: not working with rabbitmq-server 3.7 - Cannot parse invalid user line
## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: 5.3.3-1xenial
- Ruby: ruby2.4 (included in puppet package)
- Distribution: Ubuntu 16.04.3 LTS
- Module version: v8.0.0
- Rabbitmq-server : 3.7.0-1

## How to reproduce (e.g Puppet code you use)
just run
puppet agent -t

```puppet
      rabbitmq_user { 'new_user':
                admin    => true,
                password => 'pass',
                provider => 'rabbitmqctl'
        }
```

## What are you seeing
```
Error: Failed to apply catalog: Cannot parse invalid user line: warning: the VM is running with native name encoding of latin1 which may cause Elixir to malfunction as it expects utf8. Please ensure your locale is set to UTF-8 (which can be verified by running "locale" in your shell)
```
## What behaviour did you expect instead
something went wrong with name encoding when the module is trying to fetch the result of the command: 
/usr/sbin/rabbitmqctl -q list_users

the function itself is located in this file line 22:
./lib/puppet/provider/rabbitmq_user/rabbitmqctl.rb

## Output log
```
Debug: Executing: '/usr/sbin/rabbitmqctl -q list_users'
Debug: Command succeeded
Debug: Storing state
Debug: Stored state in 0.16 seconds
Error: Failed to apply catalog: Cannot parse invalid user line: warning: the VM is running with native name encoding of latin1 which may cause Elixir to malfunction as it expects utf8. Please ensure your locale is set to UTF-8 (which can be verified by running "locale" in your shell)
Debug: Dynamically-bound server lookup failed, falling back to report_server setting
```

## Any additional information you'd like to impart

### PR DESCRIPTION
**Pull request by fatmcgav**
RabbitMQ 3.7.0 seems to have changed the ordering of fields in the policies output. 

Credit to @alexizmailov for the original work in #674, I just tweaked it a bit and added some tests :) 

Fixes #671

### CODE REVIEW

File: spec/unit/puppet/provider/rabbitmq_policy/rabbitmqctl_spec.rb [hunk 1]
```diff
@@ -40,39 +40,64 @@
...
     expect { provider.exists? }.to raise_error(Puppet::Error, %r{cannot parse line from list_policies})
   end
 
-  it 'matches policies from list (>=3.2.0)' do
-    provider.class.expects(:rabbitmqctl).with('list_policies', '-q', '-p', '/').returns <<-EOT
+  context 'with RabbitMQ version >=3.7.0' do
+    it 'matches policies from list' do
+      provider.class.expects(:rabbitmqctl).with('-q', 'status').returns '{rabbit,"RabbitMQ","3.7.0"}'
+      provider.class.expects(:rabbitmqctl).with('list_policies', '-q', '-p', '/').returns <<-EOT
+/ ha-all .* all {"ha-mode":"all","ha-sync-mode":"automatic"} 0
```
Comment (wyardley) [inline spec/unit/puppet/provider/rabbitmq_policy/rabbitmqctl_spec.rb]: This mostly makes sense to me, just curious why there's a literal `.*` in the heredoc in these various examples (including in the existing code)... shouldn't the result be the literal output of running `rabbitmqctl list_policies`?

**Comment by wyardley** [comment]
Thanks for this @fatmcgav, and thanks for preserving attribution.

I'm +1 on this, with just the open question above (maybe I'm misunderstanding how the mock works here?)

My only other question is on how reliable `rabbitmq_version` is, and whether it might not make sense to just try both match patterns? But since the existing code seems to already have been using it, I guess it should work.

File: spec/unit/puppet/provider/rabbitmq_policy/rabbitmqctl_spec.rb (hunk 1 above)
Comment (fatmcgav) [reply spec/unit/puppet/provider/rabbitmq_policy/rabbitmqctl_spec.rb]: I believe `.*` can be the output from `rabbitmqctl list_polices`, as it's a regex pattern that can be used within the rabbitmq policy to control what resources the policy affects. 
Unless I'm missing your point :) 

More details here: https://www.rabbitmq.com/parameters.html#policies

Comment (wyardley) [reply spec/unit/puppet/provider/rabbitmq_policy/rabbitmqctl_spec.rb]: Got it, that makes sense then. I didn't have a rabbitmq instance handy so couldn't check the actual output, and a quick Google search failed me.

**Comment by fatmcgav** [comment]
Yeh, I'm confused by the TravisCI failures, as those tests all pass for me locally :s

**Comment by wyardley** [comment]
@fatmcgav did you try with `PUPPET_VERSION="~> 5.0"`?

**Comment by wyardley** [comment]
I'm also not seeing the error locally, even with `PUPPET_VERSION="~> 5.0"`.

I think the following should resolve the rubocop failures?

```
--- a/lib/puppet/provider/rabbitmq_policy/rabbitmqctl.rb
+++ b/lib/puppet/provider/rabbitmq_policy/rabbitmqctl.rb
@@ -20,10 +20,12 @@ Puppet::Type.type(:rabbitmq_policy).provide(:rabbitmqctl, parent: Puppet::Provid
       # / ha-all .* all {"ha-mode":"all","ha-sync-mode":"automatic"} 0 << This is for RabbitMQ v >= 3.7.0
       if Puppet::Util::Package.versioncmp(rabbitmq_version, '3.7') >= 0
         regex = %r{^(\S+)\s+(\S+)\s+(\S+)\s+(all|exchanges|queues)?\s+(\S+)\s+(\d+)$}
-        applyto_index, pattern_index = 4, 3
+        applyto_index = 4
+        pattern_index = 3
       else
         regex = %r{^(\S+)\s+(\S+)\s+(all|exchanges|queues)?\s*(\S+)\s+(\S+)\s+(\d+)$}
-        applyto_index, pattern_index = 3, 4
+        applyto_index = 3
+        pattern_index = 4
       end
 
       policy_list.split(%r{\n}).each do |line|
```

**Comment by wyardley** [comment]
Proposed a possible fix in
https://github.com/fatmcgav/puppet-rabbitmq/pull/1
Alternately, you could try dropping the context statements around the version-specific bits.
I could only replicate the test locally by switching to this pattern:
```
provider_class = Puppet::Type.type(:rabbitmq_policy).provider(:rabbitmqctl)
describe provider_class do
[...]
  let(:provider) { provider_class.new(resource) }
```

**Comment by fatmcgav** [comment]
@wyardley it looks like ure fix has got all the tests passing, so thanks for that... 😀

File: spec/unit/puppet/provider/rabbitmq_policy/rabbitmqctl_spec.rb [hunk 2]
```diff
@@ -7,11 +7,10 @@
       pattern: '.*',
       definition: {
         'ha-mode' => 'all'
-      },
-      provider: described_class.name
```
Comment (wyardley) [inline spec/unit/puppet/provider/rabbitmq_policy/rabbitmqctl_spec.rb]: wasn't 100% sure about this, but seems to work without it.

Comment (fatmcgav) [reply spec/unit/puppet/provider/rabbitmq_policy/rabbitmqctl_spec.rb]: Yeh, it does :)

**Comment by fatmcgav** [comment]
@wyardley A new release sounds good to me :)

**Comment by bastelfreak** [comment]
Thanks @fatmcgav!

**Pull request merged**